packages/buckets/src/buckets.spec.ts

```diff
@@ -271,6 +273,107 @@ describe('Buckets...', function () {
       }
     })
   })
+
+  describe('multiple pushes', function () {
+    it('should throw if roots are out of sync/order', async function () {
+      if (isBrowser) return this.skip()
+      const { root } = await client.getOrCreate('pushes')
+
+      // Grab a reference to a relatively large file
+      const pth = path.join(__dirname, '../../..', 'testdata')
+      fileSize = fs.statSync(path.join(pth, 'file1.jpg')).size
+      const stream1 = fs.createReadStream(path.join(pth, 'file1.jpg'))
+      const rootKey = root?.key || ''
+
+      // Now create a relatively small stream that should finish first
+      const stream2 = new Duplex()
+      stream2.push(Buffer.from('some small amount of text'))
+      stream2.push(Buffer.from('that spans two buffers'))
+      stream2.push(null)
+
+      // Run them in series
+      await Promise.resolve()
+        .then(() => client.pushPath(rootKey, 'dir1/file1.jpg', stream1, { root }))
+        .then(() => client.pushPath(rootKey, 'dir1/file2.txt', stream2, { root }))
+        .then(() => {
+          throw new Error('wrong error')
+        }) // This should never fire
+        // Because we should jump down to this first
+        .catch((err) => expect(err.message).to.include('non-fast-forward'))
+
+      // Confirm that our second-level dir only includes one of them
+      const { item } = await client.listPath(rootKey, '')
+      expect(item?.items[1].items).to.have.length(1)
+    })
+
+    it.skip('should not be able to push in parallel', async function () {
```

Comment:
This isn't in effect yet, but it would be super useful to have this built into buckets. I know @sanderpick already mentioned he's like to add something like this, so I have a test here ready to go.